### PR TITLE
DEV: Add schema checking to api doc testing

### PR DESCRIPTION
**Pull request by oblakeerickson**
This commit improves upon rswag which lacks schema checking. rswag
really only checks that the https status matches, but this change adds
in the json-schema_builder gem which also has schema validation.

Now we can define schemas for each of our requests/responses in the
`spec/requests/api/schemas` directory which will make our documentation
specs a lot cleaner.

If we update a serializer by either adding or removing an attribute the
tests will now fail (this is a good thing!). Also if you change the type
of an attribute say from an array to a string the tests will now fail.
This will help significantly with keeping the docs in sync with actual
code changes! Now if you change how an endpoint will respond you will
have to update the docs too in order for the tests to pass. :D

This PR is inspired by:

 https://www.tealhq.com/post/how-teal-keeps-their-api-tests-and-documentation-in-sync

### CODE REVIEW

**Comment by CvX** [comment]
`json-schema_builder` hasn't been updated in 3 years and `json-schema` gem it relies on also isn't actively maintained anymore.

We could choose a more recent alternative (e.g. [json_schemer](https://github.com/davishmcclurg/json_schemer)) to avoid having to fork and maintain those deps by ourselves down the line.

**Comment by oblakeerickson** [comment]
Yea I saw that they were pretty out of date. I didn't know about json_schemer though!!! I'll update the PR to use that, looks like it should work. Thanks!

**Comment by oblakeerickson** [comment]
Okay PR has been updated to use the newer json_schemer gem. 

It was kind of nice to use the ruby-like schema builder, but that can still be a pain to hand write as well so the prefered method of generating schema is to just take the json output of the api response and use a converter to json_schema and then to just save that .json file. Probably going forward would be to read a json_schema json file into ruby instead of writing a ruby hash for the json_schema.

**Comment by oblakeerickson** [comment]
Actually I got ahead of myself with `json_schemer`. Its not doing strict validation like `json-schema_builder` was:

```
require 'json_schemer'

schema = {
  'type' => 'object',
  'properties' => {
    'abc' => {
      'type' => 'integer',
      'minimum' => 11
    }
  }
}
schemer = JSONSchemer.schema(schema)

puts schemer.valid?({}) # true? should be false
puts schemer.valid?({ 'xyz' => 10 }) # true? should be false
```

**Comment by oblakeerickson** [comment]
I pushed another update. We can add similar strict validation by using `additionalProperties: false` and by specifying which values are required.

File: spec/requests/api/shared/shared_examples.rb [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    it "returns expected response status" do
+      expect(response.status).to eq(expected_response_status)
+    end
+  end
+
+  describe "request body" do
+    it "matches the documented request schema" do |example|
+      schemer = JSONSchemer.schema(expected_request_schema.schemer)
+      valid = schemer.valid?(params)
+      unless valid # for debugging
```
Comment (martin-brennan) [inline spec/requests/api/shared/shared_examples.rb:18]: Did you mean to leave this `unless valid` part in? If you did maybe making a helper method would be useful e.g. `expect_schema_valid(schemer, params)` that does the expectation and also the puts debugging on failure.

**Review by martin-brennan — APPROVED** [review]
Just one minor comment, you are the expert in this area and from what I can see based on the context this all looks in order :)

File: spec/requests/api/shared/shared_examples.rb (hunk 1 above)
Comment (oblakeerickson) [reply spec/requests/api/shared/shared_examples.rb:18]: Yea its handy to have when trying to figure out why a spec failed, but yes I'll work on a helper method for this

**Pull request merged**